File: 98. 验证二叉搜索树/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traversal(self, root, tlist):
        if not root:
            return
        self.traversal(root.left, tlist)
        tlist.append(root.val)
        self.traversal(root.right, tlist)
        return

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        # 利用二叉搜索树的中序遍历是递增序列的特性即可
        tlist = []
        self.traversal(root, tlist)
        for i in range(len(tlist) - 1):
            if tlist[i] >= tlist[i+1]:
                return False
        return True

File: 98. 验证二叉搜索树/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traversal(TreeNode* root, vector<int>& tlist){
        if(root == NULL) return;
        traversal(root->left, tlist);
        tlist.push_back(root->val);
        traversal(root->right, tlist);
        return;
    }
    bool isValidBST(TreeNode* root) {
        vector<int> tlist;
        traversal(root, tlist);
        for(int i = 0; i < tlist.size()-1; i++){
            if(tlist[i] >= tlist[i+1]) return false;
        }
        return true;
    }
};